### ISSUE TITLE
First roll does not return results onRollComplete

### ISSUE THREAD

**Issue by clevett** [3d-dice/dice-box#18]
I've been seeing an interesting behaviour where the first die or array of dice that I roll does not return a result onRollComplete. Every roll after works as expected. dice-box seems like its completing init fine and I have a spinner to ensure the user isn't clicking roll button too soon. You can see in the image below that I'm getting one complete message for two rolls.

![](https://i.imgur.com/6bAXdma.png)

```
 useEffect(() => {
    setIsLoading( true )
    
    const Box = new Dicebox("#dice-box", {
      assetPath: '/dice-box/',
      theme: pink,
    })

    Box.init().then(() =>  {
      setDiceBox( Box )
      setIsLoading( false )
    } );
  }, []) 
```

I tried commenting out the diceBox.add and just using the diceBox.roll. This bug was the same here.

```
const diceRoll = async ( notation: string | string[], add = false ) => {
    console.log("%c Clicked~", `color: orange; size 14px;`)

    add ? await diceBox.add( notation ) : await diceBox.roll( notation )

    if ( diceBox ) {
      diceBox.onRollComplete = (resultsArray: resultsType[]) => {
        console.log( resultsArray )
      };
    }
  }
```

**Comment by clevett** [comment]
I resolved my issue by updating the code. I recalled in our past emails that you were going to make rolls promises. 

So this worked:

```
    if ( add ) {
      await diceBox.add( notation ).then((results: resultsType[]) => {
        handleResults(results)
      })
    } else {
      await diceBox.roll( notation ).then((results: resultsType[]) => {
        handleResults(results)
      })
    }
```

Do you want me to put in a PR to update the README ? Do you intend to keep onRollComplete ?

**Comment by frankieali** [comment]
I think this is the same issue I reported in #17 
I've been reworking how the roll promises are done the last two days so they don't swallow each other. It kind of overlaps how the internal data is handled when an array of dice notations is passed in and turned into dice groups.
To summarize the issue, on .roll a promise is attached to the World class that is fulfilled onRollComplete. The problem is that every time .add is used, it hijacks the promise, essentially removing it from calls made before it. Instead of attaching this pending promise to the world class object, each .roll and .add call should have its own promise.

**Comment by frankieali** [comment]
Fixed with 0.6 release. I overhauled the way data was managed internally to resolve the promise issues.
